### ISSUE TITLE
iterating on the NDDB prompt user experience

### ISSUE THREAD

**Issue by shiftkey** [desktop/desktop#6875]
**Please describe the problem you think should be solved**

We've had the "Notification of Diverging from Default Branch" in the app for a while, but some recent changes make me think there's a better way to surface this information.

![](https://user-images.githubusercontent.com/359239/52881834-2cbea080-313c-11e9-8a4d-39b2318341e3.png)

This came up in #6862 because the user wanted a convenient way to update their fork from the upstream, and was under the impression this didn't exist in the app.

The pain points I'd like to call out:

 - seeing this prompt while you're comparing branches feels unintuitive - why is the app distracting the user from what they're trying to do?
 - the compare tab is a very narrow space, limiting what information we can show to the users
 - it's not easy to manually trigger this information, because of how we implemented it at the time

**[Optional] Do you have any potential solutions in mind?**

The blank slate improvements in #6445 have provided real-estate for showing contextual tasks, like pulling from the tracked branch:

![](https://user-images.githubusercontent.com/1322749/52889826-a88b0e00-3179-11e9-9f05-71c735bfe8fe.png)

Seeing this in action - suggesting a task to the user, and then hiding it after it's been completed - feels like a similar flow to what the NDDB prompt was trying to achieve.

How I'd iterate on this feature:

  - extract the logic for detecting and prompting the user from the compare tab and into a service that runs in the background - this gives us the ability to control when it runs, and doesn't tie it to user interactions
  - design an action UI to show when the user should merge their default branch (either from a feature branch or from the upstream of a fork) or dismiss the action (because it's not important)

We've had previous discussions about adding in additional actions (e.g. #6674) but because this action is only shown when something needs to be done I don't think it's as important to do here.

**Comment by j-f1** [comment]
I noticed this morning that making a change causes the suggestions to disappear, which makes it harder to open Atom. Maybe this blank slate should be a separate “home” screen so it’s always available? How about a new button next to the push/pull button that opens a dropdown with the available suggestions? It could even have a numeric badge for the more active suggestions (like merging the main branch or pushing to the remote).

**Comment by borpin** [comment]
> This came up in #6862 because the user wanted a convenient way to update their fork from the upstream, and was under the impression this didn't exist in the app.

I think that part of my issue is that I do not see these prompts.  I suspect this is because the repositories were cloned with Desktop before the feature was introduced.  However, I have added the upstream via the CLI.

On Github I see:
![image](https://user-images.githubusercontent.com/1322749/53129246-05623c00-355f-11e9-9ad8-950a5537e269.png)

On Desktop I do not see anything telling me there are updates
![image](https://user-images.githubusercontent.com/1322749/53129327-2c207280-355f-11e9-9cbc-c75aacb9b0c6.png)

![image](https://user-images.githubusercontent.com/1322749/53129363-3e021580-355f-11e9-93f1-78a98ae3222f.png)

![image](https://user-images.githubusercontent.com/1322749/53129395-51ad7c00-355f-11e9-8464-66cfab2d3ab9.png)

I added the upstream via the command line;

```
$ cat .git/config
[core]
        repositoryformatversion = 0
        filemode = false
        bare = false
        logallrefupdates = true
        symlinks = false
        ignorecase = true
[submodule]
        active = .
[remote "origin"]
        url = https://github.com/borpin/emoncms.git
        fetch = +refs/heads/*:refs/remotes/origin/*
[branch "master"]
        remote = origin
        merge = refs/heads/master
[remote "upstream"]
        url = https://github.com/emoncms/emoncms.git
        fetch = +refs/heads/*:refs/remotes/upstream/*
```

On making it more intuitive, why not use the UI you use on the Pull button to indicate there are updates (a number and an arrow)

![image](https://user-images.githubusercontent.com/1322749/53130294-b8339980-3561-11e9-94f6-dcfd497fc1a1.png)

on the **current repository** button (with a mouse over tool tip) like you have in the dropdown for the current repository. 
![image](https://user-images.githubusercontent.com/1322749/53130356-e87b3800-3561-11e9-8f44-66c54909a21d.png)

![image](https://user-images.githubusercontent.com/1322749/53130382-021c7f80-3562-11e9-8819-777a22f85b59.png)

At the same time, could you add the information that GitHub shows indicating what the upstream is?
![image](https://user-images.githubusercontent.com/1322749/53130433-237d6b80-3562-11e9-9121-aa91519988cc.png)

The phrase *Current Repository* is almost redundant.

**Comment by billygriffin** [comment]
TL;DR, my suggestion is that we fix the current implementation to work as expected (#5293) prior to iterating, and that will set us up nicely to roll this into a different flow (like suggested next steps) once this is functioning as intended.

I had a chat with @niik about this yesterday, and wanted to capture our discussion here for further consideration. We both think it would be an interesting use case to incorporate a nudge to update your current branch with the latest from the default branch into the "suggested next steps" flow as discussed here. However, I don't think we're there yet and I'd love to diagnose why it's not working as expected first. We've seen extremely low usage of this feature based on our metrics, and I think it's because the implementation as it exists today doesn't match what we intended. As @borpin points out above, it's not actually showing up anywhere with any level of consistency.

My understanding is that the *intended* behavior is that when there are changes on the default branch (or upstream based on the logic to determine where the reference branch is), the user sees a blue notification dot next to "History" in that tab, as shown here:

<img width="450" alt="screen shot 2019-03-06 at 2 56 16 pm" src="https://user-images.githubusercontent.com/5091167/53916682-05167600-4020-11e9-8fed-2629e6ef71fb.png">

Then, when you click on the "History" tab, you should see the notification, as shown here:

<img width="460" alt="screen shot 2019-03-06 at 2 58 09 pm" src="https://user-images.githubusercontent.com/5091167/53916797-49a21180-4020-11e9-8a8a-721cfa22d0ea.png">

However, that's not actually what's occurring, and I rarely, if ever, see it in the course of my actual work despite being frequently behind the default branch. I *think* the reproduction steps are as follows to trigger the notification:

1. Navigate to a branch that's is behind the default branch (in other words, the default branch has commits that your branch does not)
2. Click the "History" tab
3. Click into the "Select Branch to Compare" filter field and select any branch
4. Click the "Changes" tab.
5. Click the "History" tab.
6. Now you should see the notification.

There's also potentially a timing element to this where potentially a notification could be displayed *after* a merge occurs, so whatever we do here should also consider #5617. I was not able to reproduce #5573, but whatever occurs here should also check that. I'm closing both of those issues but expect that any iteration here should verify that this works as expected.

My thinking is that we ought to get this working properly as it exists today, and then evaluate whether it makes more sense to be wrapped into the "suggested next steps" view. Either way we would need to update the logic to make it work as expected, so this feels like a good first step. I'm also going to label this a bug as it's pretty clear to me that it's not working as intended.

The description of the inconsistency (that I just found after writing all this up - great 😄) is captured very nicely in #5293, which feels like a good place to do the initial work required here. And here was the original tracking issue for the feature: https://github.com/desktop/desktop/issues/4646

**Comment by borpin** [comment]
I just updated to 1.6.3 on Windows 10.

On first start, pulling down the list of repositories, there were no notifications of any repo requiring update.  Not sure exact sequence from there, but the arrow did appear eventually.

Moving to a repo that shows there are updates available, I now see a prompt to *pull* the repository.
![image](https://user-images.githubusercontent.com/1322749/53971245-6c055f00-40f4-11e9-93b2-f34266a22099.png)

but nothing in the history tab (no blue dot). When I click on the 'select branch to compare..' I see this.

![image](https://user-images.githubusercontent.com/1322749/53971463-dddda880-40f4-11e9-9e1b-063cabcb8e96.png)

**Comment by billygriffin** [comment]
@borpin Would you mind doing the repro steps I outlined in https://github.com/desktop/desktop/issues/6875#issuecomment-470302301 and verify you're able to trigger the notification, just as a sanity check?

**Comment by borpin** [comment]
> @borpin Would you mind doing the repro steps I outlined in [#6875 (comment)](https://github.com/desktop/desktop/issues/6875#issuecomment-470302301) and verify you're able to trigger the notification, just as a sanity check?

@billygriffin yes it does.

BUT (and this is where my issue started), if I am on the *master* branch (say), and this repo is a local of a fork (origin) on GitHub, I'd like a notification that tells me I am behind the *origin/master* without having to dig around for it.  I do at least know where to look (and how to do the update in Desktop rather than via CLI which is what I was doing), but there is still a 1-1 relationship that is important (to me anyway) to keep my local and fork in sync with the upstream.

**Comment by billygriffin** [comment]
@borpin Thanks! I *think* that should be the inherited behavior of the NDDB notification if it's working as intended (which it's not today). We'll look at that as well, thanks.

**Comment by rafeca** [comment]
I can try to fix this one if it's a priority.

From an initial investigation it seems like there may be two small issues working together:

1. The `AheadBehindUpdater` logic is only triggered after the user clicks on the "History" tab and on the "Select Branch" input. You can actually see how the commit differences are displayed for each branch as it gets calculated:

    ![demo](https://user-images.githubusercontent.com/408035/76964746-7bcc9b80-6923-11ea-8deb-fd0032b28ed2.gif)

   Without the `AheadBehindUpdater` logic running, the UI cannot know if the current branch is behind the upstream branch and this is why the nudge doesn't appear until opening the branch list.

2. After it's displayed, the NDDB prompt disappears after a few seconds or after performing some actions. This seems to happen because the same flag is used to show/display the blue nudge and to show/display the prompt (`isDivergingBranchBannerVisible`). While we want to hide the blue nudge just after the user clicks the "History" tab, we probably want to leave the prompt for quite longer (maybe until the user closes it?).

**Comment by guidezpl** [comment]
Would love to see this fixed since it's my only pain point! I tried @billygriffin's repro steps with a fork and did not see the notification or blue nudge. Note, the branch (and its upstream counterpart) was not named `master`.

**Comment by shiftkey** [comment]
> Would love to see this fixed since it's my only pain point! I tried @billygriffin's repro steps with a fork and did not see the notification or blue nudge.

@guidezpl I believe this has been unshipped based on https://github.com/desktop/desktop/pull/9804
